runtime/lua/vim/lsp/buf.lua

```diff
@@ -495,11 +495,17 @@ function M.outgoing_calls()
   call_hierarchy(ms.callHierarchy_outgoingCalls)
 end
 
---- @param method string
-local function type_hierarchy(method)
+--- Lists all the subtypes or supertypes of the symbol under the
+--- cursor in the |quickfix| window. If the symbol can resolve to
+--- multiple items, the user can pick one using |vim.ui.select()|.
+---@param kind "subtypes"|"supertypes"
+function M.typehierarchy(kind)
+  local method = kind == 'subtypes' and ms.typeHierarchy_subtypes or ms.typeHierarchy_supertypes
+
   --- Merge results from multiple clients into a single table. Client-ID is preserved.
   ---
   --- @param results table<integer, {error: lsp.ResponseError, result: lsp.TypeHierarchyItem[]?}>
+  --- @return [integer, lsp.TypeHierarchyItem][]
```

Comment:
Note that tuple annotations are new in 3.8.0: 

https://github.com/LuaLS/lua-language-server/blob/5d9980edae90a71842f69ec13c13ac451f64e090/changelog.md?plain=1#L5

This avoids some of the `--- @type` annotations further below.